contrib/nydus-backend-proxy/src/main.rs

```diff
@@ -0,0 +1,282 @@
+extern crate clap;
+
+use clap::{App, Arg};
+use rocket::fs::FileServer;
+use std::env;
+use std::path::PathBuf;
+
+mod blob_backend {
+    use http_range::HttpRange;
+    use nix::sys::uio;
+    use once_cell::sync::OnceCell;
+    use rocket::fs::NamedFile;
+    use rocket::http::Status;
+    use rocket::request::{self, FromRequest, Outcome};
+    use rocket::response::{self, stream::ReaderStream, Responder};
+    use rocket::Request;
+    use rocket::Response;
+    use std::collections::HashMap;
+    use std::os::unix::io::AsRawFd;
+    use std::path::{Path, PathBuf};
+    use std::{ffi, fs, io};
+
+    #[derive(Debug)]
+    struct BlobBackend {
+        root: PathBuf,
+        blobs: HashMap<ffi::OsString, fs::File>,
+    }
+    impl BlobBackend {
+        fn new(root: PathBuf) -> Self {
+            Self {
+                blobs: get_blobmap(&root),
+                root,
+            }
+        }
+    }
+    fn get_blobmap(path: &Path) -> HashMap<ffi::OsString, fs::File> {
+        let mut blobfdmap = HashMap::new();
+        for entry in path.read_dir().expect("read blobsdir failed").flatten() {
+            let filepath = entry.path();
+            if filepath.is_file() {
+                let digest = filepath.file_name().unwrap();
+                let std_file = fs::File::open(&filepath).unwrap();
+                blobfdmap.insert(digest.to_os_string(), std_file);
+            }
+        }
+        blobfdmap
+    }
+    static BLOBBACKEND: OnceCell<BlobBackend> = OnceCell::new();
+
+    pub fn init(root: PathBuf) {
+        BLOBBACKEND.set(BlobBackend::new(root)).unwrap();
+    }
+
+    #[derive(Debug)]
+    pub struct HeaderData {
+        _host: String,
+        range: String,
+    }
+
+    #[rocket::async_trait]
+    impl<'r> FromRequest<'r> for HeaderData {
+        type Error = Status;
+
+        async fn from_request(req: &'r Request<'_>) -> request::Outcome<HeaderData, Self::Error> {
+            let host = match req.headers().get_one("Host") {
+                Some(h) => h.to_string(),
+                None => "".to_string(),
+            };
+            let rangestr = match req.headers().get_one("Range") {
+                Some(h) => h.to_string(),
+                None => "".to_string(),
+            };
+            Outcome::Success(HeaderData {
+                _host: host,
+                range: rangestr,
+            })
+        }
+    }
+
+    #[rocket::head("/<_namespace>/<_repo>/blobs/<digest>")]
+    pub async fn check(
+        _namespace: PathBuf,
+        _repo: PathBuf,
+        digest: String,
+    ) -> Result<Option<FileStream>, Status> {
+        let dis = &digest[7..];
+        let blobbackend = match BLOBBACKEND.get() {
+            Some(blob) => blob,
+            None => {
+                eprintln!("BLOBBACKEND get failed!");
+                return Err(Status::InternalServerError);
+            }
+        };
+        let path = blobbackend.root.join(&dis);
+        Ok(NamedFile::open(path)
```

Comment:
The check just needs to respond with the HTTP header `Docker-Content-Digest`, why attach the file data stream?

Reply:
> The check just needs to respond with the HTTP header `Docker-Content-Digest`, why attach the file data stream?

Rocket handles `HEAD` requests by stripping the body from the response, so this file data stream will be  dropped and no reading happens. In addition, Rocket not allow to set a custom` Content-Length` and it's not currently possible to properly handle a `HEAD` request manually.